### ISSUE TITLE
Type error with partial and trio

### ISSUE THREAD

**Issue by thejcannon** [microsoft/pyright#7222]
**Describe the bug**
(with code below)

```console
% pipx run pyright /.../foo.py                                    
/.../foo.py
  /.../foo.py:17:49 - error: Argument of type "() -> list[int]" cannot be assigned to parameter "sync_fn" of type "((...) -> T@run_sync) | ((Unknown) -> T@run_sync)" in function "run_sync"
    Type "() -> list[int]" cannot be assigned to type "((...) -> Iterable[_LiteralInteger | bool]) | ((Unknown) -> Iterable[_LiteralInteger | bool])"
      Type "() -> list[int]" cannot be assigned to type "(...) -> Iterable[_LiteralInteger | bool]"
        Function return type "list[int]" is incompatible with type "Iterable[_LiteralInteger | bool]"
          "list[int]" is incompatible with "Iterable[_LiteralInteger | bool]"
            Type parameter "_T_co@Iterable" is covariant, but "int" is not a subtype of "_LiteralInteger | bool"
      Type "() -> list[int]" cannot be assigned to type "(Unknown) -> Iterable[_LiteralInteger | bool]"
        Function accepts too many positional parameters; expected 0 but received 1
          Function return type "list[int]" is incompatible with type "Iterable[_LiteralInteger | bool]"
    ... (reportArgumentType)
1 error, 0 warnings, 0 informations
% pipx run pyright --version       
pyright 1.1.350
```

**Code or Screenshots**
```python
from typing import Callable, Any
from functools import partial

import trio

def allgather_objs(obj: Any) -> list[Any]:
    ...

async def main():
    tokens: int = 0
    n_tokens: int = 1
    f: Callable[[], list[int]] = partial(
        allgather_objs,
        obj=n_tokens,
    )

    tokens += sum(await trio.to_thread.run_sync(f))
```

`trio-typing` is `==0.8.0`, see https://github.com/python-trio/trio-typing/blob/2c17e51fc33e176c4bb389c87870c4514234dbe0/trio-stubs/to_thread.pyi

**VS Code extension or command-line**
`pyright == 1.1.350`

**Comment by erictraut** [comment]
Here's a simplified example that doesn't rely on external type definitions but still exhibits the issue.

```python
from typing import Callable, Iterable, Literal, overload

async def run_sync[_T](__fn: Callable[..., _T]) -> _T: ...

@overload
def sum(__iterable: Iterable[Literal[0]]) -> int: ...
@overload
def sum[_T](__iterable: Iterable[_T]) -> _T: ...

async def main(f: Callable[[], list[int]]):
    sum(await run_sync(f))
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.352, which I just published.
